### PR TITLE
Introduce 10 of bpf tools to pmdabpf

### PR DESCRIPTION
**Pull request by smalinux**

### CODE REVIEW

File: src/pmdas/bpf/modules/exitsnoop.c [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    if(!get_item(inst, &value))
+        return PMDA_FETCH_NOVALUES;
+
+    /* bpf.exitsnoop.age */
+    if (item == AGE) {
+        atom->d = (value->event.exit_code - value->event.start_time) / 1e9;
```
Comment (andreasgerstmayr) [inline src/pmdas/bpf/modules/exitsnoop.c:419]: The calculation should include exit_**time** and not exit_**code**.
PCP includes support for different metric units and resolutions - let's use the raw resolution (nanoseconds, according to the bpf sources) and let the PCP client tools (pmval, pmrep, htop etc.) decide in which unit they want to display that metric value.
tl;dr don't divide by `1e9`, let's use the raw resolution

File: src/pmdas/bpf/modules/exitsnoop.c [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+{
+    /* bpf.exitsnoop.age */
+    metrics[AGE] = (struct pmdaMetric)
+    {
+        .m_desc = {
+            .pmid  = PMDA_PMID(cluster_id, 0),
+            .type  = PM_TYPE_DOUBLE,
+            .indom = indom_id_mapping[EXITSNOOP_INDOM],
+            .sem   = PM_SEM_INSTANT,
+            .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
```
Comment (andreasgerstmayr) [inline src/pmdas/bpf/modules/exitsnoop.c:180]: PCP metric metadata includes the metric value units. In this case we can tell the PCP client tools the metric value is in nanoseconds (`PM_TIME_NSEC`).
Example in the runqlat module: https://github.com/performancecopilot/pcp/blob/7adaaedd665d947d94b5e7acd045f7fbb53d127c/src/pmdas/bpf/modules/runqlat.c#L71
Detailed docs: https://man7.org/linux/man-pages/man3/pmLookupDesc.3.html

**Review by andreasgerstmayr — CHANGES_REQUESTED** [review]
Hi @smalinux, thanks for this PR! :tada: 

I started to review the first few tools (bashreadline, execsnoop, exitsnoop, fsslower). Looks good overall, I've added a couple of comments. I'll continue reviewing the remaining tools next week.

File: src/pmdas/bpf/modules/bashreadline.c [hunk 3]
```diff
@@ -0,0 +1,395 @@
+/*
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:1]: @andreasgerstmayr The problem with bashreadline is I can't invoke it via QA scripts, bashreadline sniff only bash command written by hands, not by scripts

[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:1]: There is a handy tool called expect(1) to simulate interactive terminal sessions, which we can use here. It supports commands like this:
```
spawn bash
send "echo this is a test from bash\r"
send "exit\r"
expect eof
```
This tells expect to start bash, send the echo command, send the exit command, and wait until bash is closed.

You can embed the above expect script in a PCP QA test like this:
```
cat <<EOF | expect
spawn bash
send "echo this is a test from bash\r"
send "exit\r"
expect eof
EOF
```

I tested it with the bashreadline bpf PMDA tool, and it works as expected - the bashreadline will have two instances, one for the echo command and one for the exit command.

[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:1]: One more note, when using the expect tool add the following line somewhere at the beginning of the QA test:
```
which expect >/dev/null 2>&1 || _notrun "expect binary not installed"
```
So the test will be skipped if the expect program is not installed.

File: src/pmdas/bpf/bpf.conf [hunk 4]
```diff
@@ -23,3 +24,118 @@ enabled = true
...
+
+# Trace out-of-memory killer events
+# See oomkill(8) man page
+#
+# Configuration options:
+# Name              - type    - default
+#
+# process_count     - int     - 20    : number of processes exec's to keep in cache
+[oomkill.so]
+enabled = true
```
[the remark this answers is not in the document]
Comment (natoscott) [reply src/pmdas/bpf/bpf.conf:45]: @andreasgerstmayr yep, makes sense to me to go for a default subset - particularly as the set is likely to grow over time.  Guess we want to consider those tools that are a/ most generally useful, b/ lowest overheads, and c/ least likely to break over time.

File: configure.ac [hunk 5]
```diff
@@ -1264,11 +1264,24 @@ if $pmda_bpf; then
     libbpf_version_minor=`echo $libbpf_version | cut -d. -f2`
     pmdabpf_modules="biolatency.so runqlat.so"
 
-    AC_MSG_CHECKING([if the bpf PMDA execsnoop module should be included])
+    AC_MSG_CHECKING([if the bpf PMDA modules should be included])
     if test "$libbpf_version_major" -eq 0 -a "$libbpf_version_minor" -lt 7; then
 	AC_MSG_RESULT([no (libbpf version required: 0.7.0, installed: $libbpf_version)])
     else
-	pmdabpf_modules="${pmdabpf_modules} execsnoop.so"
+	pmdabpf_modules="${pmdabpf_modules}\
```
[the remark this answers is not in the document]
Comment (smalinux) [reply configure.ac:1271]: All these tools (except CO-RE vfsstat) use `perf_buffer__poll` and this function is part of libbpf 0.7
Test:
```
$ git clone libbpf
$ git checkout v0.6.0
$ ack perf_buffer__poll
```

File: src/pmdas/bpf/modules/bashreadline.c [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+enum metric_indom { BASHREADLINE_INDOM, LOST_EVENTS };
+
+char* metric_names[METRIC_COUNT] = {
+    [COMM] = "bashreadline.comm",
+    [PID]  = "bashreadline.pid",
+    [LOST] = "bashreadline.lost",
+};
+
+char* metric_text_oneline[METRIC_COUNT] = {
+    [COMM] = "Command name",
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:108]: Done. but I'm not sure if I did it the right way.

[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:108]: I added LCFLAGS at the beginning of modules/GNUMakefile, which breaks trace_helpers.c!
so I want to enable std=c99 for all modules, expect *_helpers.c files...

File: src/pmdas/bpf/modules/bashreadline.c [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+    return 0;
+}
+
+static void bashreadline_register(unsigned int cluster_id, pmdaMetric *metrics, pmdaIndom *indoms)
+{
+    /* bpf.bashreadline.comm */
+    metrics[COMM] = (struct pmdaMetric)
+    {
+        .m_desc = {
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:155]: Thanks to @a3f, I got this from his code review via email. :)

File: src/pmdas/bpf/modules/execsnoop.c [hunk 8]
```diff
@@ -231,13 +231,13 @@ static void execsnoop_register(unsigned int cluster_id, pmdaMetric *metrics, pmd
...
     };
-    /* bpf.execsnoop.uid */
+    /* bpf.execsnoop.lost */
     metrics[LOST] = (struct pmdaMetric)
     {
         .m_desc = {
             .pmid  = PMDA_PMID(cluster_id, 6),
             .type  = PM_TYPE_U32,
-            .indom = indom_id_mapping[EXECSNOOP_INDOM],
+            .indom = indom_id_mapping[LOST_EVENTS],
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c:240]: I tried that and this break the bpf.execsnoop.lost metric 
```
$ git diff
diff --git a/src/pmdas/bpf/modules/execsnoop.c b/src/pmdas/bpf/modules/execsnoop.c
index 8b1b88034..e9239b6a5 100644
--- a/src/pmdas/bpf/modules/execsnoop.c
+++ b/src/pmdas/bpf/modules/execsnoop.c
@@ -237,7 +237,7 @@ static void execsnoop_register(unsigned int cluster_id, pmdaMetric *metrics, pmd
         .m_desc = {
             .pmid  = PMDA_PMID(cluster_id, 6),
             .type  = PM_TYPE_U32,
-            .indom = indom_id_mapping[LOST_EVENTS],
+            .indom = PM_INDOM_NULL,
             .sem   = PM_SEM_INSTANT,
             .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
         }
@@ -254,7 +254,7 @@ static void execsnoop_register(unsigned int cluster_id, pmdaMetric *metrics, pmd
     /* LOST_EVENTS InDom */
     indoms[1] = (struct pmdaIndom)
     {
-        indom_id_mapping[LOST_EVENTS],
+        PM_INDOM_NULL,
         1,
         execsnoop_lost,
     };

```

File: src/pmdas/bpf/modules/bashreadline.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:1]: Done. but I have a problem with installing expect(1) on my machine! 😏 so I couldn't test it for now.

File: qa/1577 [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+_prepare_pmda bpf
+trap "_cleanup; exit \$status" 0 1 2 3 15
+_stop_auto_restart pmcd
+
+# real QA test starts here
+cat <<EOF | _pmdabpf_install
+# Installed by PCP QA test $seq on `date`
+[fsslower.so]
+enabled=true
+filesystem=ext4
```
[the remark this answers is not in the document]
Comment (smalinux) [reply qa/1577:39]: Done.

File: src/pmdas/bpf/bpf.conf (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonk000) [reply src/pmdas/bpf/bpf.conf:42]: Yes, subset definitely. Primarily security concerns (eg: 'bashreadline'). Secondly, there may be some performance penalty, however I think most of the tracepoints I've seen so far _should_ be minimal overhead _most_ of the time.

File: qa/1444 [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+#
+
+seq=`basename $0`
+echo "QA output created by $seq"
+
+. ./common.bpf
+
+_pmdabpf_check
+_pmdabpf_require_kernel_version 5 0
+_pmdabpf_require_libbpf_version 0 7
```
Comment (jasonk000) [inline qa/1444:15]: I like this approach :+1:

File: src/pmdas/bpf/modules/tcpconnect.c [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+    {
+        .m_desc = {
+            .pmid  = PMDA_PMID(cluster_id, 6),
+            .type  = PM_TYPE_U32,
+            .indom = indom_id_mapping[TCPCONNECT_INDOM],
+            .sem   = PM_SEM_INSTANT,
+            .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
+        }
+    };
+    /* bpf.execsnoop.lost */
```
Comment (jasonk000) [inline src/pmdas/bpf/modules/tcpconnect.c:248]: nit: typo, `execsnoop.{uid,lost}`

File: qa/1810 [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+# real QA test starts here
+cat <<EOF | _pmdabpf_install
+# Installed by PCP QA test $seq on `date`
+[oomkill.so]
+enabled=true
+EOF
+_pmdabpf_wait_for_metric
+
+# Generate system activity for the CO-RE oomkill module
+perl -e 'while (1) { $a .= "A" x 1024; }'
```
Comment (jasonk000) [inline qa/1810:39]: on a big machine, this could take a while, are there solutions like cgroup/ulimit that could achieve it quickly?

File: qa/1795 [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+# real QA test starts here
+cat <<EOF | _pmdabpf_install
+# Installed by PCP QA test $seq on `date`
+[vfsstat.so]
+enabled=true
+EOF
+_pmdabpf_wait_for_metric
+
+# Generate system activity for the CO-RE vfsstat module
```
Comment (jasonk000) [inline qa/1795:38]: we should sleep some time here, or do something?

**Review by jasonk000 — CHANGES_REQUESTED** [review]
Great work!

I've made a few primary comment/requests for clarification. They are commented for `bashreadline` and apply to most modules.

Very excited, especially `oomkill` will be useful for us will be turned on immediately =)

File: src/pmdas/bpf/modules/bashreadline.c [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+    return false;
+}
+
+static unsigned int indom_id_mapping[INDOM_COUNT];
+
+#define METRIC_COUNT 3
+enum metric_name { COMM, PID, LOST };
+enum metric_indom { BASHREADLINE_INDOM, LOST_EVENTS };
+
+char* metric_names[METRIC_COUNT] = {
```
Comment (a3f) [inline src/pmdas/bpf/modules/bashreadline.c:100]: const char

File: src/pmdas/bpf/modules/bashreadline.c [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+    [LOST] = "Number of the lost events",
+};
+
+char* metric_text_long[METRIC_COUNT] = {
+    [COMM] = "Command name",
+    [PID]  = "Process identifier",
+    [LOST] = "Number of the lost events",
+};
+
+static unsigned int bashreadline_metric_count()
```
Comment (a3f) [inline src/pmdas/bpf/modules/bashreadline.c:118]: `(void)`

File: configure.ac (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply configure.ac]: Ok, thanks for checking. Then I only suggest to rename (checking) `if the bpf PMDA modules should be included` to (checking) `if bpf PMDA modules using perf_buffer__poll should be included`, because otherwise it sounds like no module gets included for libbpf < 0.7, which is not true (as we include biolatency and runqlat).

File: src/pmdas/bpf/bpf.conf (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/bpf.conf:42]: Jason is excited about oomkill, so I'd suggest let's start by enabling oomkill by default. It ticks the boxes for most generally useful and low overhead. It's using a kprobe though, which will break if the kernel devs decide to rename that kernel function.

There's one (minor?) issue though: we ship the same default `bpf.conf` on all systems, but some systems don't have libbpf 0.7 so we don't include oomkill there. This results in an error in the bpf PMDA log file. Should we template the config file?
```
[oomkill.so]
enabled = @BPF_PMDA_OOMKILL_ENABLED@
```
etc., like in other places?

File: src/pmdas/bpf/modules/bashreadline.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:107]: Hm, that's weird. Apparently:

* `typeof` is not part of the C99 standard (it will be a standard in C23), but it's available as a GNU extension (clang supports it as well)
* when we use `--std=c99` we disable GNU extensions [1] (therefore `typeof` doesn't work)
* when we don't use `--std=c99`, the gcc on my system (version 12.1.1) defaults to `--std=gnu17` [2], i.e. C17 with GNU extensions (therefore `typeof` works)
* however, we can use `__typeof__` instead of `typeof` and then it also works with `--std=c99`

So the irony is that by using `--std=c99` on a modern compiler we effectively downgrade the used C standard version. To fix this issue, I suggest to just change `typeof` to `__typeof__` in `trace_helpers.c`, then it compiles fine when using `LCFLAGS = $(C99_CFLAGS)` in the `GNUmakefile`.

[1] https://stackoverflow.com/a/17156740
[2] you can check this with `info gcc` --> `Standards` which reads `The default, if no C language dialect options are given, is
'-std=gnu17'.` on my system.

File: src/pmdas/bpf/modules/execsnoop.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/execsnoop.c:239]:
```
--- a/src/pmdas/bpf/modules/execsnoop.c
+++ b/src/pmdas/bpf/modules/execsnoop.c
@@ -237,7 +237,7 @@ static void execsnoop_register(unsigned int cluster_id, pmdaMetric *metrics, pmd
         .m_desc = {
             .pmid  = PMDA_PMID(cluster_id, 6),
             .type  = PM_TYPE_U32,
-            .indom = indom_id_mapping[LOST_EVENTS],
+            .indom = PM_INDOM_NULL,
             .sem   = PM_SEM_INSTANT,
             .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
         }
```
:+1:

You don't need the `indoms[1]` declaration anymore. And in `execsnoop_fetch_to_atom` currently you have
```
    if (inst == PM_IN_NULL) {
        return PM_ERR_INST;
    }
```
(i.e., return if the `inst` is `PM_IN_NULL`), but in case of the lost metric without indom, inst will be `PM_IN_NULL`. You can move
```
    /* bpf.execsnoop.lost */
    if (item == LOST) {
        atom->ul = lost_events;
        return PMDA_FETCH_STATIC;
    }
```
up before this check for example, then it will work.

File: src/pmdas/bpf/modules/bashreadline.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:1]: Let me know if you need help - I'm by no means an expert in expect(1), but maybe I can help troubleshooting the issue.

File: src/pmdas/bpf/modules/fsslower.c [hunk 16]
```diff
@@ -0,0 +1,608 @@
...
+    if (item == COMM) {
+        atom->cp = value->event.task;
+    }
+    /* bpf.fsslower.pid */
+    if (item == PID) {
+        atom->ul = value->event.pid;
+    }
+    /* bpf.fsslower.file_op */
+    if (item == T) {
+        atom->cp = &file_op[value->event.op];
```
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/fsslower.c:572]: I'm getting the following warning now:
```
fsslower.c: In function 'fsslower_fetch_to_atom':
fsslower.c:572:18: warning: assignment to 'char *' from incompatible pointer type 'char **' [-Wincompatible-pointer-types]
  572 |         atom->cp = &file_op[value->event.op];
      |                  ^
```
(it's a `char*` array now, you need to remove the `&`)

File: src/pmdas/bpf/modules/exitsnoop.c (hunk 2 above)
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/exitsnoop.c]: The second parameter to `PMDA_PMUNITS` needs to be `1` to specify that it's a time unit.

File: src/pmdas/bpf/modules/fsslower.c [hunk 17]
```diff
@@ -0,0 +1,608 @@
...
+    };
+    /* bpf.fsslower.bytes */
+    metrics[BYTES] = (struct pmdaMetric)
+    {
+        .m_desc = {
+            .pmid  = PMDA_PMID(cluster_id, 3),
+            .type  = PM_TYPE_U32,
+            .indom = indom_id_mapping[FSSLOWER_INDOM],
+            .sem   = PM_SEM_INSTANT,
+            .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
```
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/fsslower.c]: I think there's a copy&paste error, the unit of `bpf.fsslower.bytes` shouldn't be `PMDA_PMUNITS(0, 0, 0, 0, PM_TIME_NSEC, 0),`, it should be
```
PMDA_PMUNITS(1, 0, 0, PM_SPACE_BYTE, 0, 0) 
```

File: src/pmdas/bpf/modules/GNUmakefile [hunk 18]
```diff
@@ -7,19 +7,51 @@ MODULETMP = $(PCP_PMDAS_DIR)/$(IAM)/modules
...
 HELPERS_C = \
+	btf_helpers.c \
+	map_helpers.c \
+	trace_helpers.c \
+	uprobe_helpers.c \
 
 # ported header file.
 APPS_H = \
-	execsnoop.h      \
+	bashreadline.h \
```
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/GNUmakefile:30]: The configure script sets `PMDABPF_MODULES` to `runqlat.so biolatency.so` for example, and in the Makefile we have `$(PMDABPF_MODULES)` as a dependency of the `default_pcp default` targets - so only the modules listed in `PMDABPF_MODULES` are built.

The symlinks are created in any case though, but afaics that's not a problem as the files are not getting compiled.

File: qa/1444 [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+# real QA test starts here
+cat <<EOF | _pmdabpf_install
+# Installed by PCP QA test $seq on `date`
+[statsnoop.so]
+enabled=true
+EOF
+_pmdabpf_wait_for_metric
+
+# Generate system activity for the CO-RE statsnoop module
```
[the remark this answers is not in the document]
Comment (smalinux) [reply qa/1444:38]: I just added this: `touch /tmp/test-file && stat /tmp/test-file 1> /dev/null`,

File: src/pmdas/bpf/modules/bashreadline.c [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    if(!get_item(inst, &value))
+        return PMDA_FETCH_NOVALUES;
+
+    /* bpf.bashreadline.comm */
+    if (item == COMM) {
+        atom->cp = value->str_t.str;
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:369]: Since str_t.str is a fixed size array of chars, isn't it a null-terminated string?
https://github.com/performancecopilot/pcp/blob/main/vendor/github.com/iovisor/bcc/libbpf-tools/bashreadline.h#L10

File: src/pmdas/bpf/modules/execsnoop.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c:239]: Done. thanks. 
I applied this change to all other tools. 
however, still I got many instances from `bpf.bashreadline.lost` for example!

File: src/pmdas/bpf/modules/bashreadline.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (jasonk000) [reply src/pmdas/bpf/modules/bashreadline.c:369]: Good point - I guess it is in this case, I'm not sure about all cases.

File: src/pmdas/bpf/modules/bashreadline.c [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+}
+
+static char *find_readline_so()
+{
+    const char *bash_path = "/bin/bash";
+    FILE *fp;
+    off_t func_off;
+    char *line = NULL;
+    size_t line_sz = 0;
+    char path[128];
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:225]: Any suggestions for a better length? 
I got this number from `bcc/libbpf-tools/bashreadline.c`:
https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/bashreadline.c#L100

File: src/pmdas/bpf/modules/mountsnoop.c [hunk 22]
```diff
@@ -0,0 +1,640 @@
...
+            .sem   = PM_SEM_INSTANT,
+            .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
+        }
+    };
+    /* bpf.mountsnoop.mnt_ns */
+    metrics[MNT_NS] = (struct pmdaMetric)
+    {
+        .m_desc = {
+            .pmid  = PMDA_PMID(cluster_id, 6),
+            .type  = PM_TYPE_U32,
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/mountsnoop.c:298]: I think it this easy to change it to a string,
I used U32 because of this line:
https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/mountsnoop.h#L30

File: src/pmdas/bpf/modules/bashreadline.c (hunk 14 above)
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:100]: @jasonk000 So, I think I need change `char* (*metric_name_fn_t)` to `const char* (*metric_name_fn_t)`?

File: qa/1795 (hunk 13 above)
Comment (smalinux) [reply qa/1795:38]: Done! Thanks
```
$ git diff
diff --git a/qa/1795 b/qa/1795
index 44752897d..68e5c7818 100755
--- a/qa/1795
+++ b/qa/1795
@@ -35,6 +35,9 @@ enabled=true
 EOF
 _pmdabpf_wait_for_metric

+# Wait for values to appear
+pmsleep 2
+
```

File: qa/1810 (hunk 12 above)
Comment (smalinux) [reply qa/1810:39]: @andreasgerstmayr any suggestions here, please? 
My reference: https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/tools/oomkill_example.txt#L36

File: src/pmdas/bpf/modules/opensnoop.c [hunk 23]
```diff
@@ -0,0 +1,457 @@
...
+            .sem   = PM_SEM_INSTANT,
+            .units = PMDA_PMUNITS(0, 0, 0, 0, 0, 0),
+        }
+    };
+    /* bpf.opensnoop.fd */
+    metrics[FD] = (struct pmdaMetric)
+    {
+        .m_desc = {
+            .pmid  = PMDA_PMID(cluster_id, 2),
+            .type  = PM_TYPE_32,
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/opensnoop.c:202]: `fd`, `err` and `flags` are based on integer data type, so maybe PM_TYPE_32 is enough...
https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/opensnoop.h#L19-L20

File: src/pmdas/bpf/modules/mountsnoop.c [hunk 24]
```diff
@@ -0,0 +1,640 @@
...
+    if (item == LAT) {
+        atom->ull = (value->event.delta / 1000);
+    }
+    /* bpf.mountsnoop.mnt_ns */
+    if (item == MNT_NS) {
+        atom->ul = value->event.mnt_ns;
+    }
+    /* bpf.mountsnoop.fs */
+    if (item == FS) {
+        atom->cp = value->event.fs;
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/mountsnoop.c:595]: https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/mountsnoop.h#L33

File: src/pmdas/bpf/modules/tcpconnlat.c [hunk 25]
```diff
@@ -0,0 +1,458 @@
...
+    if (item == SADDR) {
+        atom->cp = strdup((char*)inet_ntop(value->event.af, &s, src, sizeof(src)));
+    }
+    /* bpf.tcpconnlat.lport */
+    if (item == LPORT) {
+        atom->ul = value->event.lport;
+    }
+    /* bpf.tcpconnlat.daddr */
+    if (item == DADDR) {
+        atom->cp = strdup((char*)inet_ntop(value->event.af, &d, dst, sizeof(dst)));
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/tcpconnlat.c:425]: Done. Thanks!

File: src/pmdas/bpf/modules/bashreadline.c [hunk 26]
```diff
@@ -0,0 +1,394 @@
...
+        return err;
+    }
+
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            handle_event, handle_lost_events, NULL, NULL);
+    if (!pb) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to open perf buffer: %d", err);
+        return err;
+    }
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:319]: Done. Thanks!

File: src/pmdas/bpf/modules/bashreadline.c [hunk 27]
```diff
@@ -0,0 +1,394 @@
...
+TAILQ_HEAD(tailhead, tailq_entry) head;
+
+static struct tailq_entry* allocElm(void)
+{
+    return malloc(sizeof(struct tailq_entry));
+}
+
+static void push(struct tailq_entry *elm)
+{
+    if (queuelength > env.process_count)
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:68]: Done. Thanks!

File: src/pmdas/bpf/modules/bashreadline.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:107]: Thanks for your time and this informative comment,
Since we are using `trace_helpers.c` from symlink, maybe It's not a good idea to patch that file? especially since this file changes quickly...

File: src/pmdas/bpf/modules/bashreadline.c [hunk 28]
```diff
@@ -0,0 +1,395 @@
...
+{
+    LIBBPF_OPTS(bpf_object_open_opts, open_opts);
+    off_t func_off;
+    int err;
+    char *val;
+
+    if ((val = pmIniFileLookup(cfg, module_name, "process_count")))
+        env.process_count = atoi(val);
+
+    if (libreadline_path) {
```
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:268]: It seems that the issue also already exists in the original tool, dito:
https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/bashreadline.c#L163-L168
Do you suggest deleting this code block? I removed it and everything works fine...

File: qa/1577 (hunk 9 above)
[the remark this answers is not in the document]
Comment (smalinux) [reply qa/1577:40]: Done. Thanks!

File: src/pmdas/bpf/modules/bashreadline.c (hunk 14 above)
Comment (jasonk000) [reply src/pmdas/bpf/modules/bashreadline.c:100]: I think this one was for @a3f ?

Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:100]: Oh, sorry ^^

File: src/pmdas/bpf/modules/bashreadline.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (smalinux) [reply src/pmdas/bpf/modules/bashreadline.c:1]: Done. Thanks!

File: qa/1810 (hunk 12 above)
Comment (andreasgerstmayr) [reply qa/1810:39]: Good question. I'm not sure about the exact algorithm the oomkiller uses when deciding which process to kill, I think it's also possible that it kills another, unrelated process (which we want to avoid). I think it's a bit risky to trigger the oomkiller in a QA test.

What do you think @natoscott @kmcdonell, should we try to trigger the oom-killer as part of a QA test?

File: src/pmdas/bpf/modules/bashreadline.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:268]: In the original tool you can set `libreadline_path`: https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/libbpf-tools/bashreadline.c#L50-L54
So I suggest to either make it also configurable, or remove this block.

File: qa/1810 (hunk 12 above)
Comment (jasonk000) [reply qa/1810:39]: Here's a cgroup example:

Single concat, runs inside 10MB
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo systemd-run --scope -p MemoryLimit=10M perl -e '$a .= "A" x 1024'
Running scope as unit: run-r43d2f5e56b8847fea964d1d1a0bc19ec.scope
[completes]
```

With a loop, is terminated:
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo systemd-run --scope -p MemoryLimit=10M perl -e 'while (1) { $a .= "A" x 1024 }'
Running scope as unit: run-r9ae1c624743547f58b4abdb773a73498.scope
Killed
```

Verifying:
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo journalctl --since -5m | grep oom-kill 
Aug 10 09:44:29 jkoch kernel: perl invoked oom-killer: gfp_mask=0xcc0(GFP_KERNEL), order=0, oom_score_adj=0
Aug 10 09:44:29 jkoch kernel: oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=run-rb41f35fa09a74858b93ba2537a1a955f.scope,mems_allowed=0,oom_memcg=/system.slice/run-rb41f35fa09a74858b93ba2537a1a955f.scope,task_memcg=/system.slice/run-rb41f35fa09a74858b93ba2537a1a955f.scope,task=perl,pid=100835,uid=0
```

Comment (natoscott) [reply qa/1810:39]: @andreasgerstmayr it does sound a little risky, and maybe also something that will be difficult to make reliable/deterministic across many kernel versions.  FWIW, there's also /proc/pid/oom_score that may help with focusing the oom-killer's attention, from memory its writable.

File: src/pmdas/bpf/modules/bashreadline.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:107]: afaics git automatically merges it, so unless there is a change in exactly that line, it shouldn't be an issue to modify this file from the remote repo.

File: src/pmdas/bpf/modules/bashreadline.c (hunk 14 above)
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/bashreadline.c:100]: Yes.

`metric_text_oneline` and `metric_text_long` could also be constants, however they get passed down to `dp->version.four.text` (in `bpf.c`) which expects a `char**`

File: src/pmdas/bpf/modules/mountsnoop.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/mountsnoop.c:595]: `event.fs` comes from `struct trace_event_raw_sys_enter *ctx` by `(const char *)ctx->args[2]`. I'd expect it to be null-terminated coming from the kernel. The upstream tool at https://github.com/iovisor/bcc/blob/a416a82/libbpf-tools/mountsnoop.c#L234 doesn't perform anything special, just prints it (where `printf` expects a null-terminated string).

**Review by andreasgerstmayr — APPROVED** [review]
Thanks again for this PR and all the updates! :100:
